ykbh/src/lib.rs

```diff
@@ -0,0 +1,283 @@
+use std::alloc::{alloc, dealloc, Layout};
+use std::convert::{TryFrom, TryInto};
+use ykpack::{
+    self, Constant, ConstantInt, IPlace, Local, LocalDecl, Statement, Terminator, TypeId,
+    UnsignedInt,
+};
+use yktrace::sir::SIR;
+
+pub struct SIRInterpreter {
+    /// Pointer to allocated memory containing a frame's locals.
+    locals: *mut u8,
+    /// The offset of each Local into locals.
+    offsets: Vec<usize>,
+    /// The layout of locals. Needed for deallocating locals upon drop.
+    layout: Layout,
+}
+
+impl SIRInterpreter {
+    pub fn new(local_decls: &Vec<LocalDecl>) -> Self {
+        // FIXME Soon this will be pre-computed and handed to us by SIR.
+        let mut offsets = Vec::new();
+        let mut layout = Layout::from_size_align(0, 1).unwrap();
+        for d in local_decls {
+            let align = SIR.ty(&d.ty).align();
+            let size = SIR.ty(&d.ty).size();
+            let l = Layout::from_size_align(size.try_into().unwrap(), align.try_into().unwrap()).unwrap();
+            let (nl, s) = layout.extend(l).unwrap();
+            offsets.push(s);
+            layout = nl;
+        }
+
+        // Allocate memory for the locals
+        let locals = unsafe { alloc(layout.pad_to_align()) };
+
+        SIRInterpreter {
+            locals,
+            offsets,
+            layout,
+        }
+    }
+
+    /// Inserts a pointer to the trace inputs into `locals`.
+    pub fn set_trace_inputs(&mut self, tio: *mut u8) {
+        // FIXME Later this also sets other already initialised variables as well as the program
+        // counter of the interpreter.
+        let ptr = self.local_ptr(&Local(1)); // The trace inputs live in $1
+        unsafe {
+            // Write the pointer value of `tio` into locals by casting it to an integer first.
+            std::ptr::write::<*mut u8>(ptr as *mut *mut u8, tio);
+        }
+    }
+
+    pub fn interpret(&mut self, body: &ykpack::Body) {
```

Comment:
```suggestion
    pub unsafe fn interpret(&mut self, body: &ykpack::Body) {
```

Given the correct `body`, this can dereference arbitrary pointers.

Reply:
Fixed in d0d1be4.